### PR TITLE
Unused variables, and typo in wildcard variable.

### PR DESCRIPTION
**Pull request by atombrella**
Not an exhaustive attempt at cleaning up (the test suite code was ignored). I can create a ticket and clean up more, if desired.

### CODE REVIEW

**Comment by timgraham** [comment]
I'm not sure if the loop variables (even if unused) must be renamed. At least the flake8 warnings must be fixed.

**Comment by atombrella** [comment]
OK. PyCharm generates an exhaustive list, I've completely ignored the unused parameter-warnings.

**Comment by timgraham** [comment]
If you want to check the tests also, that's fine. A ticket isn't needed.

File: django/http/multipartparser.py [hunk 1]
```diff
@@ -580,7 +580,7 @@ def exhaust(stream_or_iterable):
     if iterator is None:
         raise MultiPartParserError('multipartparser.exhaust() was passed a non-iterable or stream parameter')
 
-    for __ in iterator:
+    for _ in iterator:
```
Comment (vytisb) [inline django/http/multipartparser.py]: I would suggest not using `_` as a variable name as that may conflict with gettext imports.

Comment (charettes) [inline django/http/multipartparser.py]: Using `_` for unused variable is a common idiom.

Comment (atombrella) [inline django/http/multipartparser.py]: Perhaps it's useful to add a comment that `_` is gettext when `__` is used.

Comment (vytisb) [inline django/http/multipartparser.py]: I have looked through Django codebase for usages of `_` and found many places where it is used as an unused variable. So I retract my comment.
I guess a conflict is unlikely in well tested code, which Django is.

**Comment by atombrella** [comment]
I think this was the scope of what I found, and had success deleting without causing failed tests/flake8. There are some instances of passed parameters to methods which are completely unused, such as https://github.com/django/django/blob/cecc079168e8669138728d31611ff3a1e7eb3a9f/django/db/models/sql/query.py#L1109-L1116

Apart from unused variables, the test suite has a substantial amount of statements with no effect. Most of them just to trigger the evaluation of the query, I suppose.

**Comment by timgraham** [comment]
I don't see much value in changing descriptive names to `_` , even if unused.

**Comment by charettes** [comment]
> I don't see much value in changing descriptive names to _ , even if unused.

A pattern that could be used is prefixing such names with an `_` instead of completely replacing them.

**Comment by timgraham** [comment]
Unlike Python 2 (maybe that's why pycharm complains), Python 3 doesn't leak the name, so I'm not sure what justification we'd use to enforce a style like that.

Python 3:
```
>>> [i for i in [1, 2,3]]
[1, 2, 3]
>>> i
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'i' is not defined
````
Python 2:
```
>>> [i for i in [1, 2,3]]
[1, 2, 3]
>>> i
3
```

**Comment by charettes** [comment]
Prefixing unused variables with an underscore can be useful when unpacking results yield by a for loop. e.g.

```python
for _unused, used in elements:
    ...
```

**Comment by charettes** [comment]
@timgraham I agree that it doesn't make much sense in the list comprehension/generator case on Python 3.

**Comment by atombrella** [comment]
So, reverting the changes in list comprehension and generation, but keep the other changes (removal of `as e` and a few dead statements) ?

**Comment by timgraham** [comment]
Correct

File: django/http/multipartparser.py [hunk 2]
```diff
@@ -571,7 +571,6 @@ def exhaust(stream_or_iterable):
 
     Raise a MultiPartParserError if the argument is not a stream or an iterable.
     """
-    iterator = None
```
Comment (timgraham) [inline django/http/multipartparser.py]: The `if iterator is None` branch a few lines later seems unused.

Comment (atombrella) [inline django/http/multipartparser.py]: Indeed. I just removed it, and the tests are still passing :)

**Pull request merged**